### PR TITLE
refactor(NODE-5914): topology.selectServer to async-await

### PR DESCRIPTION
**Pull request by W-A-James**
#### What is changing?

- Refactored `Topology.selectServer` to async/await syntax
- Refactored `Topology.connect` to async/await syntax and add connection lock
- Refactored all references to `Topology.selectServer` in the driver including

##### Is there new documentation needed for these changes?
No

#### What is the motivation for this change?

NODE-5914

### Double check the following

- [X] Ran `npm run check:lint` script
- [X] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [X] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [X] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

File: test/unit/sdam/topology.test.js [hunk 1]
```diff
@@ -530,9 +509,7 @@ describe('Topology (unit)', function () {
         //     returning their value
         //   - verify that 10 callbacks were called
 
-        topology.connect(err => {
-          expect(err).to.not.exist;
-
+        topology.connect().then(async () => {
```
Comment (nbbeeken) [inline test/unit/sdam/topology.test.js:512]: I don't think this is the correct conversion iiuc. Before we would call finish which calls done to complete the test. But this promise is floating, the test isn't blocked on it.

File: src/sdam/topology.ts [hunk 2]
```diff
@@ -459,40 +460,44 @@ export class Topology extends TypedEventEmitter<TopologyEvents> {
...
+      if (!skipPingOnConnect && server && this.s.credentials) {
+        try {
+          await server.command(ns('admin.$cmd'), { ping: 1 }, {});
+          stateTransition(this, STATE_CONNECTED);
+          this.emit(Topology.OPEN, this);
+          this.emit(Topology.CONNECT, this);
+
+          return this;
+        } catch (error) {
+          exitWithError(error);
```
Comment (nbbeeken) [inline src/sdam/topology.ts:486]: Previously, we conditionally switched between calling a callback or emitting error dependent on whether a callback was provided. Ever since we wrapped this in promisify we were always providing a callback. Is it correct that we now always emit an error? If not then I think we can get rid of the try catch and let things throw on their own.

File: src/sdam/topology.ts [hunk 3]
```diff
@@ -964,7 +970,7 @@ function processWaitQueue(topology: Topology) {
           )
         );
       }
-      waitQueueMember.callback(e);
+      waitQueueMember.reject(e);
```
Comment (nbbeeken) [inline src/sdam/topology.ts:973]: Could we improve the variable name here? `selectorError` or something

File: src/sdam/topology.ts [hunk 4]
```diff
@@ -911,7 +917,7 @@ function drainWaitQueue(queue: List<ServerSelectionRequest>, err?: MongoDriverEr
           );
         }
       }
-      waitQueueMember.callback(err);
+      waitQueueMember.reject(err);
```
Comment (nbbeeken) [inline src/sdam/topology.ts:920]: maybe `drainError`, this is actually always going to be `MongoTopologyClosedError` at least currently, maybe something alluding to that `shutdownError`

File: src/sdam/topology.ts [hunk 5]
```diff
@@ -1027,7 +1033,7 @@ function processWaitQueue(topology: Topology) {
           )
         );
       }
-      waitQueueMember.callback(error);
+      waitQueueMember.reject(error);
```
Comment (nbbeeken) [inline src/sdam/topology.ts:1036]: Maybe `serverSelectionError`?

File: src/sdam/topology.ts [hunk 6]
```diff
@@ -105,7 +105,8 @@ export interface ServerSelectionRequest {
   mongoLogger: MongoLogger | undefined;
   transaction?: Transaction;
   startTime: number;
-  callback: ServerSelectionCallback;
+  resolve: (server: Server) => void;
+  reject: (error?: AnyError) => void;
   [kCancelled]?: boolean;
```
Comment (nbbeeken) [inline src/sdam/topology.ts:110]: I don't think this should be optional we wouldn't want a promise to reject without an error, if there's anywhere where we may call this with something nullish we should make sure 1. is that actually true or does the TS need fixing 2. if it is true we should make a runtime error that says something about the unexpected nullishness of something.

File: test/unit/assorted/server_selection_latency_window_utils.ts [hunk 7]
```diff
@@ -139,10 +138,7 @@ export async function runServerSelectionLatencyWindowTest(test: ServerSelectionL
   const selectedServers: Server[] = [];
 
   for (let i = 0; i < test.iterations; ++i) {
-    const server: Server = await promisify(topology.selectServer.bind(topology))(
-      ReadPreference.NEAREST,
-      {}
-    );
+    const server: Server = await topology.selectServer.bind(topology)(ReadPreference.NEAREST, {});
```
Comment (nbbeeken) [inline test/unit/assorted/server_selection_latency_window_utils.ts:141]: We don't need the bind here anymore :)

**Review by nbbeeken — CHANGES_REQUESTED** [review]
Some relatively small cleanups, this looks great!

**Review by nbbeeken — APPROVED** [review]

File: test/unit/sdam/topology.test.js [hunk 8]
```diff
@@ -457,30 +437,33 @@ describe('Topology (unit)', function () {
...
-          expect(requestCheck).property('callCount').to.equal(1);
+            // When server is created `connect` is called on the monitor. When server selection
+            // occurs `requestCheck` will be called for an immediate check.
+            expect(requestCheck).property('callCount').to.equal(1);
 
-          topology.close();
-          done();
-        });
+            topology.close();
+            done();
```
Comment (nbbeeken) [inline test/unit/sdam/topology.test.js:465]: Still using done here we should switch to returing the promise. This test is crashing the unit runs because the call count is 5 🤔

**Review by nbbeeken — CHANGES_REQUESTED** [review]

**Review by nbbeeken — APPROVED** [review]

File: test/unit/sdam/topology.test.js [hunk 9]
```diff
@@ -450,38 +430,47 @@ describe('Topology (unit)', function () {
...
+            expect(err).to.exist;
+            expect(err).to.match(/Server selection timed out/);
+            expect(err).to.have.property('reason');
 
-          // When server is created `connect` is called on the monitor. When server selection
-          // occurs `requestCheck` will be called for an immediate check.
-          expect(requestCheck).property('callCount').to.equal(1);
+            // When server is created `connect` is called on the monitor. When server selection
+            // occurs `requestCheck` will be called for an immediate check.
+            expect(requestCheck).property('callCount').to.equal(1);
```
Comment (nbbeeken) [inline test/unit/sdam/topology.test.js:466]: Test failing, change req for viz.

**Review by nbbeeken — CHANGES_REQUESTED** [review]

**Review by nbbeeken — APPROVED** [review]

File: test/unit/sdam/topology.test.ts [hunk 10]
```diff
@@ -442,86 +408,81 @@ describe('Topology (unit)', function () {
...
+      // When server is created `connect` is called on the monitor. When server selection
+      // occurs `requestCheck` will be called for an immediate check.
+      expect(requestCheck).to.have.been.calledOnce;
+      topology.close();
     });
 
-    it('should disallow selection when the topology is explicitly closed', function (done) {
-      const topology = topologyWithPlaceholderClient('someserver:27019');
-      this.sinon.stub(Server.prototype, 'connect').callsFake(function () {
+    it('should disallow selection when the topology is explicitly closed', async function () {
```
Comment (nbbeeken) [inline test/unit/sdam/topology.test.ts:445]: We are so close, 🤏🏻  unit failure here.

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/sdam/topology.ts [hunk 11]
```diff
@@ -411,17 +402,26 @@ export class Topology extends TypedEventEmitter<TopologyEvents> {
...
+      return this.connectionLock;
+    }
+
+    try {
+      this.connectionLock = this._connect(options);
+      await this.connectionLock;
+    } finally {
+      // release
+      this.connectionLock = undefined;
+    }
```
Comment (baileympearson) [inline src/sdam/topology.ts:416]: (optional) thoughts?

```suggestion
    this.connectionLock ??= this._connect(options);
    try {
    	await this.connectionLock;
    	return this;
    } finally {
    	this.connectionLock = undefined;
    }
```

File: src/sdam/topology.ts [hunk 12]
```diff
@@ -883,7 +877,7 @@ function updateServers(topology: Topology, incomingServerDescription?: ServerDes
   }
 }
 
-function drainWaitQueue(queue: List<ServerSelectionRequest>, err?: MongoDriverError) {
+function drainWaitQueue(queue: List<ServerSelectionRequest>, drainError: MongoDriverError) {
```
Comment (baileympearson) [inline src/sdam/topology.ts:880]: 🔥 I always appreciate drive-by improvement of variable names, thanks for improving this one!

lack of attention to variable names leads to death of readability by a thousand cuts 🔪

File: test/unit/assorted/server_selection_spec_helper.js [hunk 13]
```diff
@@ -102,104 +102,84 @@ function executeServerSelectionTest(testDefinition, testDone) {
...
+        continue;
       }
+      expect(selectedServerDescription[prop]).to.deep.equal(
+        value,
+        `Mismatched selected server "${prop}"`
+      );
+    }
+    return;
+  } catch (err) {
+    // are we expecting an error?
```
Comment (baileympearson) [inline test/unit/assorted/server_selection_spec_helper.js:174]: tbh I thought this was intended to be a question for the reviewers 😅 

I don't feel strongly but maybe making it a statement instead of a question would be a little more clear

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: src/sdam/topology.ts (hunk 11 above)
Comment (nbbeeken) [reply src/sdam/topology.ts:416]:
```suggestion
    try {
      return await (this.connectionLock ??= this._connect(options));
    } finally {
      this.connectionLock = undefined;
    }
```

Let's play ⛳ (kidding... mostly 👀)

Comment (baileympearson) [reply src/sdam/topology.ts:416]:
```suggestion
    return await (this.connectionLock ??= this._connect(options)).finally(() => this.connectionLock = undefined);
```

😤

**Review by nbbeeken — CHANGES_REQUESTED** [review]
just a formatting lint 🎉

**Review by nbbeeken — APPROVED** [review]

**Review by baileympearson — APPROVED** [review]
